### DISCUSSION TITLE
Absolute humidity and mixing ratio added to global variables and/or temp/hum sensors?

### DISCUSSION THREAD

**Discussion by pkkrusty** [arendst/Tasmota#14234] [Ideas]
As absolute humidity can be calculated with just temperature and relative humidity, and mixing ratio with pressure, does anyone want to do some work integrating those very useful measurements into the Tasmota system? I've taken a stab at the formulas in the support.ino file by copying the dew point function, but that's only a small part of the work required, and I don't know the ecosystem to make it all work. There was some work on this a few years ago in #5579 but doesn't look like anyone ran with it.

Functions below (also not sure how exponents work in C++):

```
float CalcAbsHum(float t, float h)
{
  if (isnan(h) || isnan(t)) { return NAN; }

  if (Settings->flag.temperature_conversion) {                 // SetOption8 - Switch between Celsius or Fahrenheit
    t = (t - 32) / 1.8;                                       // Celsius
  }

// formula from https://www.hatchability.com/Vaisala.pdf - Simplified equation with 0.083% max error from -20 - 50ºC

  float result = (((6.116441*10^(7.591386*t/(t+240.7263)))*h/100)/(t+273.15))*2.16679*100;

  return result;  // units are g/m^3
}

float CalcMixingRatio(float t, float h, float p)
{
  if (isnan(h) || isnan(t)) { return NAN; }

  if (Settings->flag.temperature_conversion) {                 // SetOption8 - Switch between Celsius or Fahrenheit
    t = (t - 32) / 1.8;                                       // Celsius
  }

  if (isnan(p)) {
    p = 1013.25;                                              // Average pressure in hectopascals at sea level
  }
  
  // formula from https://www.hatchability.com/Vaisala.pdf - Simplified equation with 0.083% max error from -20 - 50ºC

  float result = 6.116441*10^(7.591386*t/(t+240.7263))*621.9907/(p-(6.116441*10^(7.591386*t/(t+240.7263))))*h/100;

  return result; // units are g/kg and will generally range from 0-90g within the accurate range of the formula
}
```

**Comment by boozeman** [comment]
Hi,

When using ESP32 we can also use [Berry](https://tasmota.github.io/docs/Berry/) (and I Love it!!) for this. My example setup has BME280 and SHT31 sensors and some greenhouse related parameters are calculated with BME280 Pressure and SHT31 Temp and Humidity. The example is my first Berry driver, so it might look horrible, but it makes what I am expecting.

[GHData.be](https://gist.github.com/boozeman/e191e01ef914e2651544a03a7acf32eb)

**Reply by Haussuche** [reply]
Love the idea

**Reply by sfromis** [reply]
As things are now, the drivers for SGP30, SGP40 and SEN5X already provides absolute humidity, but it is indeed not a complex calculation based on temperature and relative humidity.

**Reply by Haussuche** [reply]
That is great, but are they available for all Humidity sensors?

**Reply by sfromis** [reply]
If they were, I'd not have stated it for only 3 specific sensor drivers.
